src/Components/Ignitor/src/CancellableOperation.cs

```diff
@@ -5,37 +5,50 @@
 using System.Threading;
 using System.Threading.Tasks;
 
+#nullable enable
 namespace Ignitor
 {
     internal class CancellableOperation<TResult>
     {
-        public CancellableOperation(TimeSpan? timeout)
+        public CancellableOperation(TimeSpan? timeout, CancellationToken cancellationToken)
         {
             Timeout = timeout;
 
             Completion = new TaskCompletionSource<TResult>(TaskContinuationOptions.RunContinuationsAsynchronously);
             Completion.Task.ContinueWith(
                 (task, state) =>
                 {
-                    var operation = (CancellableOperation<TResult>)state;
+                    var operation = (CancellableOperation<TResult>)state!;
                     operation.Dispose();
                 },
                 this,
                 TaskContinuationOptions.ExecuteSynchronously); // We need to execute synchronously to clean-up before anything else continues
 
+
+            Cancellation = CancellationTokenSource.CreateLinkedTokenSource(cancellationToken);
+
             if (Timeout != null && Timeout != System.Threading.Timeout.InfiniteTimeSpan && Timeout != TimeSpan.MaxValue)
             {
-                Cancellation = new CancellationTokenSource(Timeout.Value);
-                CancellationRegistration = Cancellation.Token.Register(
-                    (self) =>
-                    {
-                        var operation = (CancellableOperation<TResult>)self;
-                        operation.Completion.TrySetCanceled(operation.Cancellation.Token);
-                        operation.Cancellation.Dispose();
-                        operation.CancellationRegistration.Dispose();
-                    },
-                    this);
+                Cancellation.CancelAfter(Timeout.Value);
             }
+
+            CancellationRegistration = Cancellation.Token.Register(
+                (self) =>
+                {
+                    var operation = (CancellableOperation<TResult>)self!;
+
+                    if (cancellationToken.IsCancellationRequested)
```

Comment:
All this work to use the object passed in and it still makes a closure here 😄

Reply:
Commenting on a merge PR really helps nobody 🥇